src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala

```diff
@@ -252,37 +274,63 @@ private[sql] class OapFileFormat extends FileFormat
           }
 
-          version match {
-            case DataFileVersion.OAP_DATAFILE_V1 =>
-              val reader = new OapDataReaderV1(file.filePath, m, partitionSchema, requiredSchema,
-                filterScanners, requiredIds, pushed, oapMetrics, conf, enableVectorizedReader,
-                options, filters, context = context)
-              reader.read(file)
-            // Actually it shouldn't get to this line, because unsupported version will cause
-            // exception thrown in readVersion call
-            case _ =>
-              throw new OapException("Unexpected data file version")
-              Iterator.empty
+          var orcWithEmptyColIds = false
+          // if enableVectorizedReader == true, init VectorizedContext,
+          // else context is None.
+          var context: Option[VectorizedContext] = None
+          if (enableVectorizedReader) {
+            if (isParquet) {
+              context = Some(ParquetVectorizedContext(partitionSchema,
+                file.partitionValues, returningBatch))
+            } else if (isOrc) {
+              val readerOptions = OrcFile.readerOptions(conf).filesystem(fs)
+              val reader = OrcFile.createReader(path, readerOptions)
+              val requestedColIdsOrEmptyFile = OrcUtils.requestedColumnIds(
+                isCaseSensitive, dataSchema, requiredSchema, reader, conf)
+              if (!requestedColIdsOrEmptyFile.isEmpty) {
+                val requestedColIds = requestedColIdsOrEmptyFile.get
+                assert(requestedColIds.length == requiredSchema.length,
+                  "[BUG] requested column IDs do not match required schema")
+                context = Some(OrcVectorizedContext(partitionSchema, file.partitionValues,
+                  returningBatch, requiredSchema, dataSchema, enableOffHeapColumnVector,
+                    copyToSpark, requestedColIds))
+              } else {
+                orcWithEmptyColIds = true
+                context = None
+              }
+            }
+          } else {
+            context = None
           }
 
+          if (orcWithEmptyColIds) {
+            // For the case of Orc with empty required column Ids.
+            Iterator.empty
+          } else {
+            version match {
+              case DataFileVersion.OAP_DATAFILE_V1 =>
+                val reader = new OapDataReaderV1(file.filePath, m, partitionSchema, requiredSchema,
+                  filterScanners, requiredIds, pushed, oapMetrics, conf, enableVectorizedReader,
+                  options, filters, context)
+                reader.read(file)
+              // Actually it shouldn't get to this line, because unsupported version will cause
+              // exception thrown in readVersion call
+              case _ =>
+                throw new OapException("Unexpected data file version")
+                Iterator.empty
+            }
+          }
```

Comment:
I saw these same changes already in the previous pull request (/pull/3). Suggest to avoid.

Reply:
This is my fault. I made a mistake that this PR has two commits which includes the third PR. Thanks for the reminder.